Src/Sarif.PatternMatcher.Sdk/ValidatorBase.cs

```diff
@@ -118,18 +118,75 @@ public static bool IsBase64EncodingOfPrintableCharacters(string secret)
             return true;
         }
 
-        public static bool ContainsDigitAndChar(string matchedPattern)
+        public static bool ContainsDigitAndChar(string secret)
         {
-            bool oneDigit = false, oneLetter = false;
+            return ContainAtLeastNDigit(secret, 1) && ContainAtLeastNLetter(secret, 1);
+        }
+
+        public static bool ContainLowerAndUpperCaseLetter(string secret)
+        {
+            return ContainAtLeastNLowerCaseLetter(secret, 1) && ContainAtLeastNUpperCaseLetter(secret, 1);
+        }
+
+        public static bool ContainAtLeastNDigit(string secret, int n)
+        {
+            int digitCount = 0;
 
-            foreach (char ch in matchedPattern)
+            foreach (char c in secret)
             {
-                if (char.IsDigit(ch)) { oneDigit = true; }
-                if (char.IsLetter(ch)) { oneLetter = true; }
-                if (oneDigit && oneLetter) { return true; }
+                if (char.IsDigit(c))
+                {
+                    digitCount++;
+                }
             }
 
-            return false;
+            return digitCount >= n;
```

Comment:
>return digitCount >= n [](http://example.com/codeflow?start=12&length=22)

I would add this check inside the if statement to short circuit once the number of chars is found.